From 0a02ab5badaa2cdc7aadd02f1eb3c75dad14734f Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Mon, 8 Sep 2014 15:07:19 -0400
Subject: [ethernet] ixgbevf: Adds function to set PSRTYPE register

Message-id: <1410188883-6595-5-git-send-email-jogreene@redhat.com>
Patchwork-id: 92330
O-Subject: [RHEL7.1 PATCH 04/48] ixgbevf: Adds function to set PSRTYPE register
Bugzilla: 1091124
RH-Acked-by: Neil Horman <nhorman@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: John Linville <linville@redhat.com>

BZ: 1091124
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7919168
Tested: In beaker. See BZ for details

This patch creates a new function to set PSRTYPE. This function helps lay
the ground work for eventual multi queue support.

Signed-off-by: Alexander Duyck <alexander.h.duyck@intel.com>
Signed-off-by: Don Skidmore <donald.c.skidmore@intel.com>
Tested-by: Phil Schmitt <phillip.j.schmitt@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
(cherry picked from commit 1bb9c6390e0f6df355576b48514b393a2579c3b9)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/ixgbevf/ixgbevf_main.c b/drivers/net/ethernet/intel/ixgbevf/ixgbevf_main.c
index e2f10af..144668d 100644
--- a/drivers/net/ethernet/intel/ixgbevf/ixgbevf_main.c
+++ b/drivers/net/ethernet/intel/ixgbevf/ixgbevf_main.c
@@ -1107,6 +1107,21 @@ static void ixgbevf_configure_srrctl(struct ixgbevf_adapter *adapter, int index)
  IXGBE_WRITE_REG(hw, IXGBE_VFSRRCTL(index), srrctl);
 }
 
+static void ixgbevf_setup_psrtype(struct ixgbevf_adapter *adapter)
+{
+ struct ixgbe_hw *hw = &adapter->hw;
+
+ /* PSRTYPE must be initialized in 82599 */
+ u32 psrtype = IXGBE_PSRTYPE_TCPHDR | IXGBE_PSRTYPE_UDPHDR |
+        IXGBE_PSRTYPE_IPV4HDR | IXGBE_PSRTYPE_IPV6HDR |
+        IXGBE_PSRTYPE_L2HDR;
+
+ if (adapter->num_rx_queues > 1)
+  psrtype |= 1 << 29;
+
+ IXGBE_WRITE_REG(hw, IXGBE_VFPSRTYPE, psrtype);
+}
+
 static void ixgbevf_set_rx_buffer_len(struct ixgbevf_adapter *adapter)
 {
  struct ixgbe_hw *hw = &adapter->hw;
@@ -1154,8 +1169,7 @@ static void ixgbevf_configure_rx(struct ixgbevf_adapter *adapter)
  int i, j;
  u32 rdlen;
 
- /* PSRTYPE must be initialized in 82599 */
- IXGBE_WRITE_REG(hw, IXGBE_VFPSRTYPE, 0);
+ ixgbevf_setup_psrtype(adapter);
 
  /* set_rx_buffer_len must be called before ring initialization */
  ixgbevf_set_rx_buffer_len(adapter);
-- 
1.7.1